contrib/epee/include/syncobj.h

```diff
@@ -150,6 +154,127 @@ namespace epee
   };
 
 
+  struct reader_writer_lock {
+
+    reader_writer_lock() noexcept : 
+    write_mode(false),
+    read_mode(0) {};
+
+    bool start_read() noexcept {
```

Comment:
That is the core problem of what makes this complex. `Blockchain` APIs in `blockchain.cpp` can call each other recursively. 

Now you can redesign all of them, and remove recursive locking. But that would be much more work than this. And needs a lot of logical/semantical changes.

Reply:
Then follow the existing naming scheme for `boost::scoped_lock` and `boost::shared_lock`.

Reply:
What naming scheme? 

And keep in mind, these four methods: 

1. `start_read`
2. `end_read`
3. `start_write`
4. `end_write`

Should not be used directly.  The main interface to use this lock is `RLOCK` and `RWLOCK`. The only reason we need those 4 methods is in case the user wants to do something custom with the lock. For example in this specific case [1] we have to use these methods instead of `RLOCK` and `RWLOCK` since we are locking and unlocking the lock inside the scope [2]. 

1. https://github.com/0xFFFC0000/monero/blob/494d6be27e9de576f291428d42fc401d4196327f/src/cryptonote_core/blockchain.cpp#L5045
2. https://github.com/0xFFFC0000/monero/blob/494d6be27e9de576f291428d42fc401d4196327f/src/cryptonote_core/blockchain.cpp#L5065

Reply:
> What naming scheme?

If you used the names that `shared_mutex` uses, then `unique_lock` and `shared_lock` could be used instead of the `RLOCK` and `RWLOCK` macros. This does not prevent direct calls into those functions.

So:
 * `start_read` -> `lock_shared`
 * `end_read` -> `unlock_shared`
 * `start_write` -> `lock`
 * `end_write` -> `unlock`

Then `RLOCK` becomes `boost::shared_lock<reader_writer_lock>` and `RWLOCK` becomes `boost::unique_lock<reader_writer_lock>`. I think this is preferable; no allocation is performed.

However, using those classes may require other functions (such as `try_shared_lock`)?

Reply:
I see your point. Few points to consider: 

1. `RLOCK`/`RWLOCK` was a @moneromooo-monero  suggestion and appeared to be a good solution. 
2. What you are saying makes sense to me too. But We have to go one level deeper. And possibility introduces a lot of complexity where we exactly don't want any complexity. For example, we have to evaluate what are the **_exact_** requirements of `boost::shared_lock` and `boost::unique_lock` [1].  That includes support for all the operations they provide. On top of that, that code should be debugged and make sure 100% of those features (that we don't want 99% of them) don't have any bugs or side effects. That looks to me like a slippery slope. 




https://github.com/boostorg/thread/blob/aec18d337f41d8e3081ee65f5cf3b5090179ab0e/include/boost/thread/lock_types.hpp#L503

Reply:
Keep in mind, that we eventually want to remove recursive locking. This PR is a mid-term solution. Until we clean all the APIs inside the `blockchain`. Once all of those APIs are cleaned. We can simply use `std::shared_mutex`  and `std::mutex`.